odoo/service/server.py

```diff
@@ -117,6 +125,55 @@ def server_activate(self):
         if not self.reload_socket:
             super(ThreadedWSGIServerReloadable, self).server_activate()
 
+    def process_request(self, request, client_address):
+        """
+        Start a new thread to process the request.
+        Override the default method of class socketserver.ThreadingMixIn
+        to be able to get the thread object which is instantiated and maintain information about it.
+        """
+        t = threading.Thread(target = self.process_request_thread,
+                             args = (request, client_address))
+        t.daemon = self.daemon_threads
+        t.start()
+        self._request_threads[t] = {'start_time': time.time()}
+
+    def _handle_request_noblock(self):
+        """
+        In `socketserver`,
+        when `shutdown` is called
+        (e.g. below in this file `httpd.shutdown()`)
+        `__shutdown_request` is set to True,
+        https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L241-L249
+        This tells the server to no longer accept requests. Callstack below:
+        - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L231-L234
+        - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L308
+        - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L487
+
+        Unfortunately, this is very possible the HTTP server is already polling for a request when this flag is set
+        to `True` by the main thread:
+        - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L232
+        As the poll_interval is set to 0.5s:
+        - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L215
+        if a request arrives within 500ms after the server is shut down,
+        the request is accepted, while the server is shutdown.
+        We want to prevent this: Once the server is marked as shut down, we want it to no longer accept requests:
+        so the inverse proxy knows it has to queue the request.
+        """
+        if self._BaseServer__shutdown_request:
+            return
+        super(ThreadedWSGIServerReloadable, self)._handle_request_noblock()
+
+    @property
+    def request_threads(self):
+        """
+        A property returning the alive request threads.
+        """
```

Comment:
Why a proprety ?

Reply:
I can do a method if you prefer, it was just to make it more elegant when I had to use loop on the threads, but it does make the code a bit more complicated.